rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,207 @@
+
+## Proposal
+
+Assets should be included individually on a per-component basis and included in the component partial.
+
+This would mean that a page loads `component.css` and `component.js` only when a component is present on the page - rather than the current behaviour, which sees component assets loaded if used anywhere within an application. If multiple uses of the same component occur on the page, there should only be one instance of the assets included on the page.
+
+### Keep the stylesheets in the head
+
+Placing stylesheets in the `head` will ensure that all of the required CSS is downloaded and parsed before rendering begins. This is the current behaviour, so shouldn't be changed.
+
+Whilst putting a `link rel="stylesheet"` into the body is [okay][link_stylesheet_is_body_okay], this would lead to too many changes at once, too many unknowns, and isn't considered best practice. If this proposal is implemented, further investigation could be carried to find out which is the better way of loading assets to improve painting performance.
+
+### Keep JavaScript files just before the closing body tag
+
+[Slimmer][slimmer] currently takes all `script` elements that don't have `defer` or `async` attributes and [moves them to the end of the markup][slimmer_moves_scripts]. This ensures that they are parsed and executed in a known order when the DOM is ready to be read. This behaviour shouldn't be changed.
+
+### Shared assets
+
+The assets should be placed in a folder that allows each application to reference them. This feature is due as part of the replatforming work - but that should not block the serving of individual assets as there are performance advantages that don't rely on having a shared folder.
+
+A version identifier should be added to the component asset filename to help identify which version of the components gem is being used - for example `gem-v29-2-1-accordion-09288...adb36.css`. The fingerprint should remain to show that the contents of the file are the same.
+
+The bespoke app assets should be renamed to their application name - so collections would have `collections-5f801d...179b0.css` and `collections-828a8...abb6c.js`. This will make it easier to see which stylesheet comes from which application.
+
+Because each application is using the components gem as the source for the assets for each component, the compiled assets will be the same - whether the resulting CSS or JavaScript file is produced by whitehall, frontend, or collections. When writing to the shared folder the component assets may be overwritten, but - thanks to the fingerprint - they'll be overwritten with files containing the same content.
+
+## Considerations
```

Comment:
I think that this should be an optional choice and backwards compatible - so if that means running a `accordion-individual.scss` file that `@import`s from `govuk_frontend_support` and from the `_accordion.scss` then that's okay. I'm aware that there are applications that don't want this approach, and the gem needs to work (and work without taking a long time) for both publishing and rendering applications.

A quick check of frontend running using app-live in GOVUK Docker, cold start:

* using gem: 17 seconds
* using local version of the gem: 44 seconds
* using local version of the gem with `@import govuk_frontend_support` for each component that requires it: 59 seconds

Whilst not great, I think it's an acceptable increase whilst there are other avenues worth investigating to make this faster - off the top of my head, more efficient ways of only importing the specific things that are needed from GOVUK Frontend, using the faster Dart Sass, use of `@use` within the components gem and rendering applications.

This is only a single run, so I'll do some more investigation into this tomorrow to be able to report an average number of runs. 

Whilst not a complete solution, running it using `./startup.sh --live` seems to be about four times faster, which makes me suspicious that I'm not doing a completely cold start.

I'd suggest that an unacceptable regression would be if a cold start is above a minute under the current libsass and Docker set up.

Reply:
Great. Yup having files without needing the import sounds great to avoid performance penalties everywhere. If it's a 25% increase as per your experiment that isn't too bad either - it's an order of magnitude better than I feared.

A minute sounds good too, so long as we have a certain point where we're prepared to go "eep maybe we should focus on getting dart in first" works for me

Yeah to do a full cold start you need to clear the cache, `rails assets:clobber tmp:clear` I think to be complete